include/onnxruntime/core/session/onnxruntime_cxx_api.h

```diff
@@ -1049,43 +1121,230 @@ struct ArenaCfg : Base<OrtArenaCfg> {
+  Unowned<Value> GetOutput(size_t index, const int64_t* dim_values, size_t dim_count) const;
+  Unowned<Value> GetOutput(size_t index, const std::vector<int64_t>& dims) const;
+  void* GetGPUComputeStream() const;
+
+ private:
+  OrtKernelContext* ctx_;
+};
+
+/// <summary>
+/// This struct owns the OrtKernInfo* pointer when a copy is made.
+/// For convenient wrapping of OrtKernelInfo* passed to kernel constructor
+/// and query attributes, warp the pointer with Ort::Unowned<KernelInfo> instance
+/// so it does not destroy the pointer the kernel does not own.
+/// </summary>
+struct KernelInfo : Base<OrtKernelInfo> {
+  explicit KernelInfo(std::nullptr_t) {} ///< Create an empty instance to initialize later
+  explicit KernelInfo(OrtKernelInfo* info); ///< Take ownership of the instance
+  KernelInfo Copy() const;
+  
+  template <typename T>  // T is only implemented for float, int64_t, and string
+  T GetAttribute(const char* name) const;
+
+  template <typename T>  // T is only implemented for std::vector<float>, std::vector<int64_t>
+  std::vector<T>  GetAttributes(const char* name) const;
+};
+
+/// <summary>
+/// Create and own custom defined operation.
+/// </summary>
+struct Op : Base<OrtOp> {
+  explicit Op(std::nullptr_t) {}  ///< Create an empty Operator object, must be assigned a valid one to be used
+  
+  explicit Op(OrtOp*);  ///< Take ownership of the OrtOp
+  
+  static Op Create(const OrtKernelInfo* info, const char* op_name, const char* domain,
+                  int version, const char** type_constraint_names,
+                  const ONNXTensorElementDataType* type_constraint_values,
+                  size_t type_constraint_count,
+                  const OpAttr* attr_values,
+                  size_t attr_count,
+                  size_t input_count, size_t output_count);
+
+  void Invoke(const OrtKernelContext* context,
+              const Value* input_values,
+              size_t input_count,
+              Value* output_values,
+              size_t output_count);
+
+  // For easier refactoring
+  void Invoke(const OrtKernelContext* context,
+              const OrtValue* const* input_values,
+              size_t input_count,
+              OrtValue* const* output_values,
+              size_t output_count);
+};
+
+/// <summary>
+/// This entire structure is deprecated, but we not marking
+/// it as a whole yet since we want to preserve for the next release.
+/// </summary>
 struct CustomOpApi {
   CustomOpApi(const OrtApi& api) : api_(api) {}
 
-  template <typename T>  // T is only implemented for std::vector<float>, std::vector<int64_t>, float, int64_t, and string
-  T KernelInfoGetAttribute(_In_ const OrtKernelInfo* info, _In_ const char* name);
-
+  /** \deprecated use Ort::Value::GetTensorTypeAndShape()
+   * [[deprecated]]
+   * This interface produces a pointer that must be released. Not exception safe.
+   */
+  [[deprecated("use Ort::Value::GetTensorTypeAndShape()")]]
```

Comment:
I've been trying to use these new APIs in custom ops (e.g., onnxruntime-extensions), and they fail because `GetApi()` returns `nullptr`. 

Please correct me if I'm mistaken, but custom op dlls may not have access to `OrtGetApiBase` (undefined symbol), which is needed by `InitApi()` to initialize the global api object returned by `GetApi()`. Instead, custom ops rely on being passed a pointer to `OrtApiBase` (via the exported `RegisterCustomOps` function) from which they can access the OrtApi object.

Assuming that the above is correct, what is the plan to ensure that APIs like `Ort::Value::GetTensorTypeAndShape()` work in such environments? Perhaps we need a way to construct these Ort::XXX classes with an existing OrtApi object? I ask because I'm also trying to write my own KernelInfo APIs outside of this soon-to-be deprecated CustomOpApi struct. 

Thanks!

Reply:
I should mention that my _temporary workaround_ is to create the overload `InitApi(const OrtApi* api)`, which initializes the global api. I currently call this function on entry to `RegisterCustomOps`.

Reply:
I think we should follow the official way for now.

Reply:
What is the official way for custom ops like those in onnxruntime-extensions? Calling the current version of `InitApi()` does not compile/link for the reasons outlined above.

Should custom ops manually set the api global as below?
```c++
#define ORT_API_MANUAL_INIT
#include <onnxruntime_cxx_api.h>
#undef ORT_API_MANUAL_INIT

OrtStatus* ORT_API_CALL RegisterCustomOps(OrtSessionOptions* options, const OrtApiBase* api_base) {
  Global<void>::api_ = api_base->GetApi(ORT_API_VERSION);
  // ...
}
```

If so, creating a _new_ overload of `InitApi()` (as I did above) may be nicer:
```c++
OrtStatus* ORT_API_CALL RegisterCustomOps(OrtSessionOptions* options, const OrtApiBase* api_base) {
  Ort::InitApi(api_base->GetApi(ORT_API_VERSION)); // This overload is new and does not exist yet.
  // ...
}
```

Or, do you mean that custom ops should use the deprecated CustomOpApi for now?

Reply:
I feel we should make this a subject of a separate PR since I am already making too many changes and we have time until next release?